### DISCUSSION TITLE
URL proxy rewrites for multi-tenancy

### DISCUSSION THREAD

**Discussion by dotnize** [solidjs/solid-start#1376] [Ideas]
It would be nice to have a feature similar to [NextResponse.rewrite()](https://nextjs.org/docs/app/api-reference/functions/next-response#rewrite) in Next.js, allowing domain-based routing and proxying. This would enable different route structures for specific domains/tenants without the need for multiple apps or deployments.

An example of this is the [Vercel Platforms Starter Kit](https://github.com/vercel/platforms) ([middleware.ts](https://github.com/vercel/platforms/blob/main/middleware.ts#L72)) which has 2 sets of routes:
1. `app/app` for the root domain (app.vercel.pub)
2. `app/[domain]` for all other tenants (subdomains and custom domains)

SvelteKit also has a [`reroute`](https://kit.svelte.dev/docs/hooks#universal-hooks-reroute) (https://github.com/sveltejs/kit/pull/11537) hook that probably makes this possible, but I couldn't find an example since it was just recently released in January.

Here are other proposals/RFCs from other frameworks that I've found:

Astro: https://github.com/withastro/roadmap/discussions/665, https://github.com/withastro/roadmap/discussions/681
Remix: https://github.com/remix-run/remix/discussions/2857, https://github.com/remix-run/remix/discussions/2886

Would the team be interested in exploring this feature?
Are there potential approaches or existing patterns that can be used as a workaround? (besides using an external/third-party proxy or multiple deployments)

**Update: already possible - see https://egoist.dev/multi-tenant-web-app-2024#solid-start**

**Comment by FlatMapIO** [comment]
I have the same requirement, currently using a CF worker proxy to rewrite URLs is very uncomfortable, and I will encounter some stream interruption related issues

**Comment by ryansolid** [comment]
I imagine we can already rewrite from middleware on server requests.. But if I understand this correctly we need to do so on client routing as well. So this is a router feature isn't it? Like a way of intercepting the path before sending it to matching?

**Reply by dotnize** [reply]
Yes, but it seems like this is already possible with a few modifications!
https://egoist.dev/multi-tenant-web-app-2024#solid-start

**Reply by Sleepful** [reply]
https://archive.md/FYnXW

**Reply by Sleepful** [reply]
Ryan explains how the FileRouter works here, `35:41`:

https://www.youtube.com/live/r147AHzbys0&t=2141

Which is relevant to the blog article about multi-tenant fileRouter
